pubsub.go

```diff
@@ -259,28 +267,53 @@ func (p *PubSub) processLoop(ctx context.Context) {
 		p.peers = nil
 		p.topics = nil
 	}()
+
 	for {
 		select {
-		case s := <-p.newPeers:
-			pid := s.Conn().RemotePeer()
-			ch, ok := p.peers[pid]
+		case pid := <-p.newPeers:
+			_, ok := p.peers[pid]
 			if ok {
-				log.Error("already have connection to peer: ", pid)
-				close(ch)
+				log.Warning("already have connection to peer: ", pid)
+				continue
 			}
 
 			messages := make(chan *RPC, 32)
-			go p.handleSendingMessages(ctx, s, messages)
 			messages <- p.getHelloPacket()
-
+			go p.handleNewPeer(ctx, pid, messages)
 			p.peers[pid] = messages
 
+		case s := <-p.newPeerStream:
+			pid := s.Conn().RemotePeer()
+
+			_, ok := p.peers[pid]
+			if !ok {
+				log.Warning("new stream for unknown peer: ", pid)
+				s.Reset()
+				continue
+			}
+
 			p.rt.AddPeer(pid, s.Protocol())
 
+		case pid := <-p.newPeerError:
+			delete(p.peers, pid)
+
 		case pid := <-p.peerDead:
 			ch, ok := p.peers[pid]
-			if ok {
-				close(ch)
+			if !ok {
+				continue
+			}
+
+			close(ch)
+
+			if p.host.Network().Connectedness(pid) == inet.Connected {
+				// still connected, must be a duplicate connection being closed.
+				// we respawn the writer as we need to ensure there is a stream active
+				log.Warning("peer declared dead but still connected; respawning writer: ", pid)
```

Comment:
There's a chance we'll hit this frequently the connection may close before we realize we're disconnected. It may not be ab issue but we should watch for this.